src/ctapipe/io/datawriter.py

```diff
@@ -297,50 +289,180 @@ def __enter__(self):
             self._writer.write(
                 table_name="simulation/event/subarray/shower",
                 containers=[event.index, event.simulation.shower],
             )
 
-            for tel_id, sim in event.simulation.tel.items():
-                table_name = self.table_name(tel_id)
-                tel_index = _get_tel_index(event, tel_id)
-                self._writer.write(
-                    f"simulation/event/telescope/impact/{table_name}",
-                    [tel_index, sim.impact],
-                )
+    def _write_dl0_subarray(self, event):
+        self._writer.write(
+            table_name="dl0/event/subarray/trigger",
+            containers=[event.index, event.dl0.trigger],
+        )
 
-        if self.write_r1_waveforms:
-            self._write_r1_telescope_events(event)
+    def _write_telescope_event(self, tel_event):
+        table_name = self.table_name(tel_event.index.tel_id)
+        self._writer.write(
+            f"simulation/event/telescope/impact/{table_name}",
+            [tel_event.index, tel_event.simulation.impact],
+        )
+
+        if tel_event.simulation is not None:
+            self._write_simulation_telescope(tel_event)
+            self._write_constant_pointing(tel_event)
 
         if self.write_r0_waveforms:
-            self._write_r0_telescope_events(event)
+            self._write_r0_telescope(tel_event)
+
+        if self.write_r1_waveforms:
+            self._write_r1_telescope(tel_event)
 
-        # write telescope event data
-        self._write_dl1_telescope_events(event)
+        # always write dl0, contains needed trigger info etc.
+        self._write_dl0_telescope(tel_event)
+
+        if self.write_dl1_images or self.write_dl1_parameters:
+            self._write_dl1_telescope(tel_event)
 
-        # write DL2 info if requested
         if self.write_dl2:
-            self._write_dl2_telescope_events(event)
-            self._write_dl2_stereo_event(event)
+            self._write_dl2_telescope(tel_event)
 
         if self.write_muon_parameters:
-            self._write_muon_telescope_events(event)
+            self._write_muon_telescope(tel_event)
+
+    def _write_r0_telescope(self, tel_event: TelescopeEventContainer):
+        table_name = self.table_name(tel_event.index.tel_id)
+        tel_event.r0.prefix = ""
+        self._writer.write(
+            f"r0/event/telescope/{table_name}",
+            [tel_event.index, tel_event.r0],
+        )
+
+    def _write_r1_telescope(self, tel_event: TelescopeEventContainer):
+        table_name = self.table_name(tel_event.index.tel_id)
+        tel_event.r1.prefix = ""
+        self._writer.write(
+            f"r1/event/telescope/{table_name}",
+            [tel_event.index, tel_event.r1],
+        )
+
+    def _write_dl0_telescope(self, tel_event: TelescopeEventContainer):
+        self._writer.write(
+            "dl0/event/telescope/trigger",
```

Comment:
Shouldn't there be a `{table_name}` here? Maybe I'm misunderstanding what the `_telescope` part of the function name implies.

Reply:
This is an open issue, see https://github.com/cta-observatory/ctapipe/issues/2139.

I didn't change the data format here, so the trigger table doesn't have a `{table_name}`